netket/driver/qsr.py

```diff
@@ -0,0 +1,384 @@
+        _mels[last_i: last_i + len_i] = mels[start_i:end_i]
+
+        last_i = last_i + len_i
+        _secs[n + 1] = last_i
+
+        _maxlen = max(_maxlen, len_i)
+
+    padding_factor = max(
+        MAX_LEN, min_padding_factor
+    )  # minimum padding at 64 to avoid excessive recmpilation
+    padded_size = padding_factor * int(np.ceil(last_i / padding_factor))
+
+    _sigma_p = _sigma_p[:padded_size, :]
+    _mels = _mels[:padded_size]
+
+    return _sigma_p, _mels, _secs, _maxlen
+
+
+@partial(jax.jit, static_argnums=0)
+def _avg_O(afun, pars, model_state, sigma):
+    sigma = sigma.reshape((-1, sigma.shape[-1]))
+    _, vjp = nkjax.vjp(lambda W: afun(
+        {"params": W, **model_state}, sigma), pars)
+    (O_avg,) = vjp(jnp.ones(sigma.shape[0]) / sigma.shape[0])
+    return jax.tree_map(lambda x: mpi.mpi_mean_jax(x)[0], O_avg)
+
+
+####
+
+
+@partial(jax.vmap, in_axes=(None, None, 0, 0, None), out_axes=0)
+def map_masked_offset(fun, arr, offset, size, MAX_SIZE):
+    masked_fun = jax.mask(fun, ("l",), "")
+    return masked_fun((jax.lax.dynamic_slice(arr, [offset], [MAX_SIZE]),), dict(l=size))
+
+
+def sum_sections(arr, secs, MAX_SIZE):
+    """
+    Equivalent to
+    for i in range(secs.size-1):
+        out[i] = jnp.sum(arr[secs[i]:secs[i+1]])
+
+    secs must be 1 longer than the number of sections desired.
+    MAX_SIZE is the max length of a section
+    """
+    offsets = secs[:-1]
+    sizes = secs[1:] - offsets
+    return map_masked_offset(jnp.sum, arr, offsets, sizes, MAX_SIZE)
+
+
+@partial(jax.jit, static_argnums=(0, 5))
+def local_value_rotated_kernel(log_psi, pars, sigma_p, mel, secs, MAX_LEN):
+    log_psi_sigma_p = log_psi(pars, sigma_p)
+    U_sigma_sigma_p_psi_sigma_p = mel * jnp.exp(log_psi_sigma_p)
+
+    return jnp.log(sum_sections(U_sigma_sigma_p_psi_sigma_p, secs, MAX_LEN))
+
+
+@partial(jax.jit, static_argnums=(0, 6))
+def grad_local_value_rotated(log_psi, pars, model_state, sigma_p, mel, secs, MAX_LEN):
+    log_val_rotated, vjp = nkjax.vjp(
+        lambda W: local_value_rotated_kernel(
+            log_psi, {"params": W, **model_state}, sigma_p, mel, secs, MAX_LEN
+        ),
+        pars,
+    )
+    log_val_rotated, _ = mpi.mpi_mean_jax(log_val_rotated)
+
+    (O_avg,) = vjp(jnp.ones_like(log_val_rotated) / log_val_rotated.size)
+
+    O_avg = jax.tree_map(lambda x: mpi.mpi_mean_jax(x)[0], O_avg)
+
+    return log_val_rotated, O_avg
+
+
+@jax.jit
+def compose_grads(grad_neg, grad_pos):
+    return jax.tree_util.tree_multimap(
+        lambda x, y: 2.0 * jnp.conj(x - y),
+        grad_neg,
+        grad_pos,
+    )
+
+
+# for nll
+@partial(jax.jit, static_argnums=(0, 5))
+def local_value_rotated_amplitude(log_psi, pars, sigma_p, mel, secs, MAX_LEN):
+    log_psi_sigma_p = log_psi(pars, sigma_p)
+    U_sigma_sigma_p_psi_sigma_p = mel * jnp.exp(log_psi_sigma_p)
+
+    return jnp.log(jnp.abs(sum_sections(U_sigma_sigma_p_psi_sigma_p, secs, MAX_LEN)) ** 2)
+
+
+#####
+
+
+@jax.jit
+def tree_norm(a):
```

Comment:
move this to jax utils?

Reply:
I would not move this to `nk.jax`. This is computing the L2 norm that is correct if your tree is representing a vector. But trees sometimes represent matrices. So `norm` is a bit a generic term.

Eventually we will discuss how to do this properly, but for now I'd avoid defining APIs that might bite us in the future.

It's a one-liner, and the only reason I had defined it is because I wanted to make the code more explicit.